### PR TITLE
Support quoted literal search terms for CLI flags (e.g. --dry-run) (#14041)

### PR DESCRIPTION
**Pull request by chishxd**
## Purpose

Fixes search for CLI-flag-style terms (e.g. --dry-run, -v) which are currently
unsearchable due to two separate issues in the search pipeline:

1. Indexer : `_word_re` in `sphinx/search/__init__.py` only matched `\w+`,
   so hyphens were treated as word separators at index-build time. A page mentioning
   `--dry-run` would index the term as `dry` and `run` — the flag itself was never
   stored as a searchable token.

2. Query parser : `searchtools.js`'s `splitQuery`/`_parseQuery` treats any
   term beginning with `-` as an exclusion operator. Even if the index preserved the
   token, searching `--dry-run` (or `-v`) would silently exclude results rather than
   search for them.

The simple fix:

- `_word_re` is updated to (?<!\w)-{1,2}\w[\w-]*|\w+ which  preserves leading
single/double hyphens as part of a token only at a word boundary

- `splitQuery` now extracts "quoted phrases" from the query string before running
the normal tokenizer on the remainder, returning {quotedTerms, plainTerms}
instead of a flat array. _parseQuery adds quoted terms directly to searchTerms,
bypassing the - exclusion check entirely

**Scope / known limitations:**
- This does not change behavior for *unquoted* CLI-flag searches... typing `--dry-run`
  without quotes still goes through the normal splitter/exclusion path. Quoting is
  the documented workaround, matching the direction discussed in the linked issue.
- `SearchLanguage.js_splitter_code`, the existing hook that lets custom language
  search implementations override `splitQuery`, is not yet updated to the new
  `{quotedTerms, plainTerms}` return shape. A custom override returning a flat array
  would silently break with this change. I haven't touched other-language search
  implementations, flagging this for maintainer input on whether it needs handling
  in this PR or a follow-up.

Tests added:
- Python: unit tests for `_word_re` covering CLI flags, short flags, and the
  compound-word regression guard.
- JS: unit tests for `splitQuery` (quoted extraction, unchanged plain-term behavior)
  and `_parseQuery` (quoted terms bypass exclusion).
- Integration: a `sphinx-build` fixture asserting `--dry-run` appears as an indexed
  term in the built `searchindex.js`, and that a quoted search for it returns results.

`CHANGES.rst` entry added.

## References

- Fixes #14041
- Related: #13892

## AI Disclosure

AI tools were used during preparation of this PR:

-  I used Claude for design discussion and planning.. mostly like a rubber duck, working through the
  distinction between the indexer-side and query-parser-side problems, reviewing the
  regex design for edge cases (compound words, word-boundary detection), and
  reviewing/debugging the JS diff stuff
- Also used GLM with opencode for test suite... 

All code in this PR was reviewed and understood by me before submission, and all
interaction with maintainers/reviewers on this PR will be done by me directly, per
the project's AI policy.

### CODE REVIEW

**Comment by chishxd** [comment]
woahh that's some crazy amount of Red crosses... But i also tested running test on clean master branch, and got these same errors, the Pygments update seems to have broken test assertions... And Maybe I messed up some linting error, for which I will push some update soon

**Comment by Ashark** [comment]
@AA-Turner Hello! Could you please take a look at this MR?  
I really really want this, for kde-builder documentation.

File: tests/test_search.py [hunk 1]
```diff
@@ -514,3 +514,83 @@ def test_check_js_search_indexes(make_app, sphinx_test_tempdir, directory):
         f'Search index fixture {existing_searchindex} does not match regenerated copy.'
     )
     assert fresh_searchindex.read_bytes() == existing_searchindex.read_bytes(), msg
+
+
+def test_word_re_tokenizes_cli_flags() -> None:
```
Comment (wlach) [inline tests/test_search.py]: I would love it if these were parameterized test cases

File: tests/test_search.py [hunk 2]
```diff
@@ -514,3 +514,83 @@ def test_check_js_search_indexes(make_app, sphinx_test_tempdir, directory):
...
+    # Verify the indexed terms contain document references
+    # Terms can be int (single doc) or list (multiple docs)
+    run_term = index['terms']['run']
+    verbos_term = index['terms']['verbos']
+    v_term = index['terms']['-v']
+
+    assert (isinstance(run_term, int) and run_term > 0) or (
+        isinstance(run_term, list) and len(run_term) > 0
+    )
+    assert (isinstance(verbos_term, int) and verbos_term > 0) or (
```
Comment (wlach) [inline tests/test_search.py]: verbose?

File: CHANGES.rst [hunk 3]
```diff
@@ -13,6 +13,10 @@ Bugs fixed
   English stemmer) and Dutch (which uses the Dutch Porter stemmer).
   Patch by Hugo van Kemenade
 
+* #14041: Fix CLI flags like ``--dry-run`` and ``-v`` being unsearchable due to
+  word splitter regex and query parser behavior changes.
+  Patch by User
```
Comment (wlach) [inline CHANGES.rst]: fill this in?

File: sphinx/search/__init__.py [hunk 4]
```diff
@@ -85,7 +87,7 @@ class SearchLanguage:
 };
 """
 
-    _word_re = re.compile(r'\w+')
+    _word_re = re.compile(r'(?<!\w)-{1,2}\w[\w-]*|\w+')
```
Comment (wlach) [inline sphinx/search/__init__.py:94]: maybe add a quick comment explaining the complexity

File: sphinx/search/__init__.py [hunk 5]
```diff
@@ -118,7 +120,9 @@ def word_filter(self, word: str) -> bool:
 
 
 # SearchEnglish imported after SearchLanguage is defined due to circular import
-from sphinx.search.en import SearchEnglish  # NoQA: E402
+from sphinx.search.en import (  # ruff: ignore[module-import-not-at-top-of-file]
```
Comment (wlach) [inline sphinx/search/__init__.py]: why these changes?

**Review by wlach — COMMENTED** [review]
I can't approve PRs but fwiw here's my feedback. In general looks like a reasonable change to me

File: tests/test_search.py (hunk 1 above)
Comment (chishxd) [reply tests/test_search.py]: Oh Okay.. will make it parameterized in a Commit.

File: CHANGES.rst (hunk 3 above)
Comment (chishxd) [reply CHANGES.rst]: Oh.. Forgot about it I guess..

File: tests/test_search.py (hunk 2 above)
Comment (chishxd) [reply tests/test_search.py]: oh I jst let it be verbos because tokenization makes is verbos at the end

File: sphinx/search/__init__.py (hunk 5 above)
Comment (chishxd) [reply sphinx/search/__init__.py]: New ruff version seems to be using a different type of comment.. I will revert it back asw 😄
